controller/controller_client.go

```diff
@@ -5,284 +5,251 @@
+	state, _, _ := unit.Expected()
+	c.units.SetOutput(unit)
+	if state == client.UnitStateHealthy { // config update, so restart
+		_ = unit.UpdateState(client.UnitStateStopping, "shipper is restarting", nil)
+		c.stopShipper()
+		c.startShipper(unit)
+	} else if state == client.UnitStateStopped { // shut down
+		_ = unit.UpdateState(client.UnitStateStopping, "shipper is stopping", nil)
+		c.stopShipper()
+		_ = unit.UpdateState(client.UnitStateStopped, "shipper is stopped", nil)
+	}
+}
 
-	// reset the local logger
-	c.log = logp.L()
-
-	err = c.Run(cfg, unit)
+/*
+//////////
+////////// Input Handlers
+/////////
+*/
+
+func (c *clientHandler) startgRPC(unit *client.Unit, cfg config.ShipperClientConfig) {
+	//TODO: until we get TLS config fixed/figured out, run in insecure mode
+	// certPool := x509.NewCertPool()
+	// for _, cert := range cfg.Shipper.Server.TLS.CAs {
+	// 	if ok := certPool.AppendCertsFromPEM([]byte(cert)); !ok {
+	// 		c.reportError("error appending cert obtained from input in shipper startup", err, outUnit)
+	// 		return
+	// 	}
+	// }
+
+	creds := insecure.NewCredentials() //:= credentials.NewTLS(&tls.Config{
+	// 	ClientAuth:     tls.RequireAndVerifyClientCert,
+	// 	ClientCAs:      certPool,
+	// 	GetCertificate: c.getCertificate,
+	// 	MinVersion:     tls.VersionTLS12,
+	// })
+
+	err := c.grpcServer.Start(creds, cfg.Server)
 	if err != nil {
-		c.log.Errorf("error running shipper: %s", err)
-		_ = unit.UpdateState(client.UnitStateFailed, err.Error(), nil)
+		c.reportError("failed to start grpc server", err, unit)
 		return
 	}
+	c.log.Debugf("gRPC started")
+
+}
+
+func (c *clientHandler) stopGRPC() {
+	c.grpcServer.Stop()
+	c.log.Debugf("gRPC server stopped")
 }
 
 // start an individual input stream
-func (c *clientHandler) startStream(unit *client.Unit) {
-	// when we have individual input streams, that'll go here.
-	c.log.Debugf("Got unit stream for processor: %s", unit.ID())
+func (c *clientHandler) addInput(unit *client.Unit) {
+	// I assume that once we connect output processors, that will go somewhere here
+	_, _, cfg := unit.Expected()
+	// decode the gRPC config used by the shipper
+	conn := config.ShipperClientConfig{}
+	cfgObj, err := libcfg.NewConfigFrom(cfg.Source.AsMap())
+	if err != nil {
+		c.reportError("error creating config object", err, unit)
+		return
+	}
+	err = cfgObj.Unpack(&conn)
+	if err != nil {
+		c.reportError("error unpacking input config", err, unit)
+		return
+	}
+	c.units.AddUnit(unit, conn)
+	c.log.Debugf("Got client %s with config: Server: %s", unit.ID(), conn.Server)
 	_ = unit.UpdateState(client.UnitStateHealthy, "healthy", nil)
 
+	// figure out if we need to initialize the gRPC endpoint
+	// TODO: this is another thing that will change with https://github.com/elastic/elastic-agent-shipper/issues/225,
+	// As we'll have a dedicated unit for the input, and we won't be using random input unit updates to see if we need to update the gRPC endpoint.
+	c.startgRPC(unit, conn)
```

Comment:
This assumes that the connection information cannot change once the process is started. Is that actually true?

This eventually calls `c.grpcServer.Start(creds, cfg.Server)` which just does nothing if the server is already started but doesn't consider that the configuration could hav e changed.

Could the agent decide to change the port we listen on for example?

Reply:
So, my understanding after talking with @blakerouse is that the connection is static across the lifetime of the shipper.

A complicating factor here is the TLS settings, which are currently unimplemented but tracked in https://github.com/elastic/elastic-agent-shipper/issues/224; I'm not sure if those could change over the lifetime of the shipper, but @leehinman might know?

Reply:
They could change over the lifetime of the shipper.  If for example a CA or server certificate expires and needs to be updated.  But I think in general, both the server side and output side should expect to be restarted independently.